### PR TITLE
DPP-723 Add global parallelism limit for ACS queries

### PR DESCRIPTION
**Pull request by rautenrieth-da**
This PR limits the number of ACS contract id fetching queries running in parallel (across all ACS streams).

# Performance impact

I have measured the performance impact using the following method:
- Ingest 1M contracts, where 50% contracts are of template `Foo1` and 50% of those contracts are immediately archived
- Start 16 concurrent ACS streams for `Foo1`
- Measure average items/sec per stream

The results don't show any improvement of limiting the parallelism. Variance between different runs with the same settings is quite high (up to 20% difference).

```
Parallelism   Items/second
1             1538,68
2             2683,97
4             4169,93
8             4323,48
16            4277,05
32            4062,43
64            4238,05
```

The same for 64 concurrent streams:
```
2    608,61
4    944,82
8    1047,07
1024 968,88
```

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    while (running < parallelism && waiting.nonEmpty) {
+      val head = waiting.dequeue()
+      startTask(
+        head.task().andThen { case result => head.promise.tryComplete(result) }(executionContext)
+      )
+    }
+  }
+}
+
+class ThreadPoolBasedConcurrencyLimiter[T](
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:58]: Remove?

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  override def execute(task: () => Future[T]): Future[T] = task()
+}
+
+class QueueBasedConcurrencyLimiter[T](
+    parallelism: Int,
+    executionContext: ExecutionContext,
+) extends ConcurrencyLimiter[T] {
+  assert(parallelism > 0)
+
+  case class WaitingTask(
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:25]: You do not really need so much data to track. We could as good as do line 44 and line 52 as we compose the future generating function at line 34. You do not need to have a future generating function neither, just something to create the future and being done with it. So the queue-s signature can be () => Unit.

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      task: () => Future[T],
+      promise: Promise[T] = Promise[T](),
+  )
+  private val waiting = mutable.Queue[WaitingTask]()
+  private var running: Int = 0
+
+  override def execute(task: () => Future[T]): Future[T] = synchronized {
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:32]: If we do not care about T in the queue, then we could easily move the generic parameter T from the class to the method. This makes it more generic, and limiting would not relate to what we are limiting.

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    if (running >= parallelism) {
+      val waitingTask = WaitingTask(task)
+      waiting.enqueue(waitingTask)
+      waitingTask.promise.future
+    } else {
+      startTask(task())
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:38]: I would consider not doing this optimization with the enqueue/start, and just having one path with enqueueing, which would make the code lighter

**Review by nmarton-da — COMMENTED** [review]
Looks good!
As discussed offline I would not go with a complete threadpool for this purpose, the performance difference is minimal.
Nice test suite.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala [hunk 5]
```diff
@@ -670,6 +671,8 @@ private class JdbcLedgerDao(
         acsFetchingparallelism = acsContractFetchingParallelism,
         metrics = metrics,
         materializer = materializer,
+        // TODO: properly parametrize the parallelism. Currently using 8, because the default connection pool size is 16.
```
[the remark this answers is not in the document]
Comment (rautenrieth-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala]: Added a `--acs-global-parallelism-limit` CLI option for this, defaulting to 10.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala [hunk 6]
```diff
@@ -69,15 +71,18 @@ class FilterTableACSReader(
...
-        dispatcher
-          .executeSql(metrics.daml.index.db.getActiveContractIds)(
-            eventStorageBackend.activeContractEventIds(
-              partyFilter = query.filter.party,
-              templateIdFilter = query.filter.templateId,
-              startExclusive = query.fromExclusiveEventSeqId,
-              endInclusive = activeAt._2,
-              limit = idPageSize,
-            )
+        idQuerylimiter
```
[the remark this answers is not in the document]
Comment (rautenrieth-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala]: Moved the limiting to `eventStorageBackend.activeContractEventBatch`

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala (hunk 2 above)
Comment (rautenrieth-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala]: Refactored the code to make the class non-parametric.

**Review by kamil-da — APPROVED** [review]
LGTM!

**Comment by nmarton-da** [comment]
> This PR limits the number of ACS contract id fetching queries running in parallel (across all ACS streams).
> 
> I have tried to measure the performance impact using the following method:
> 
>     * Ingest 1M contracts, where 50% contracts are of template `Foo1` and 50% of those contracts are immediately archived
> 
>     * Start 16 concurrent ACS streams for `Foo1`
> 
>     * Measure average items/sec per stream
> 
> 
> However, the differences between different parallelism limits were the same order of magnitude as the variance between runs with the same settings.

does these still apply after fixing the configuration flag?

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    result
+  }
+
+  private def enqueueTask[T](task: () => Future[T]): Future[T] = synchronized {
+    val promise = Promise[T]()
+
+    // Note: tasks are started sequentially (see startTasks), but they can finish concurrently.
+    // Only the block withing `andThen` needs to be synchronized.
+    val waitingTask = () => {
+      running = running + 1
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:39]: would you consider moving this to startTasks? Looks so dangerous this state modification in a lambda without synchronized ;)

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    parallelism: Int,
+    executionContext: ExecutionContext,
+) extends ConcurrencyLimiter {
+  assert(parallelism > 0)
+
+  type Task = () => Unit
+  private val waiting = mutable.Queue[Task]()
+  private var running: Int = 0
+
+  override def execute[T](task: () => Future[T]): Future[T] = synchronized {
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:27]: enqueueTask with a startTask() as the before last statement would give this method. maybe just inline?

File: ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala (hunk 7 above)
Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala:39]: (also would be more natural to make the incrementing next to the while)

**Review by nmarton-da — APPROVED** [review]
LGTM!
Left some minor comments on the logic.
Would be nice to see how this behaves, and fine tuning that magic 10 number, but that can be a different PR cc @kamil-da

**Comment by nmarton-da** [comment]
> This PR limits the number of ACS contract id fetching queries running in parallel (across all ACS streams).
> # Performance impact
> 
> I have measured the performance impact using the following method:
> 
>     * Ingest 1M contracts, where 50% contracts are of template `Foo1` and 50% of those contracts are immediately archived
> 
>     * Start 16 concurrent ACS streams for `Foo1`
> 
>     * Measure average items/sec per stream
> 
> 
> The results don't show any improvement of limiting the parallelism. Variance between different runs with the same settings is quite high (up to 20% difference).
> 
> ```
> Parallelism   Items/second
> 1             1538,68
> 2             2683,97
> 4             4169,93
> 8             4323,48
> 16            4277,05
> 32            4062,43
> 64            4238,05
> ```
> 
> The same for 64 concurrent streams:
> 
> ```
> 2    608,61
> 4    944,82
> 8    1047,07
> 1024 968,88
> ```

Thanks for testing, it is clearly visible it works :+1: 
I wouldn't worry too much about the impact as you go higher: it seem to me from your experiments, that something else is the bottle-neck in those cases (probably CPU? 1000 * 64 = 4000 * 16). Also test scenario gives ideal situation for postgresql to cache, and this limiting is primarily for limiting impact on disk.
Feel free to merge, LGTM

**Pull request merged**